### PR TITLE
Refactoring weave test helpers

### LINKED ISSUE #372: Create a package for testing tools
This is the result of https://github.com/iov-one/weave/pull/364#discussion_r discussion 

Create a package that will contain functions and types helpful during tests. 

This is a migration/cleanup for the [`x/helpers.go`](https://github.com/iov-one/weave/blob/master/x/helpers.go) file. Instead of using the [`TestHelpers`](https://github.com/iov-one/weave/blob/master/x/helpers.go#L15) as a namespace we want to use a package.  Code will be moved gradually.

## Acceptance criteria

- [ ] pick a good name (as opposed to [bad names](https://blog.golang.org/package-names#TOC_5.)) for the package,
- [ ] As a first step move handlers implementations only from the `x/helpers.go` to the new package. 
- [ ] Replace the use of [a custom handler mock in `x/cash` tests](https://github.com/iov-one/weave/blob/master/x/cash/dynamicfee_test.go#L287) with the one from the new package

### PR DESCRIPTION
**Pull request by husio**
A new package `weavetest` is created. It contains all the commonly used
functions and structures for tests.
All one time used mocks were moved next to the test using them.

This is a YUGE change. We did everything for this sprint so I hope this is not an issue. It is also not that hard to follow on each test helper as it goes instead of only moving `helpers.go` and adjust the import paths.

resolve #372 and much more! :muscle: :dragon: 

Next step is to go through all the tests once more and extract/refactor more if needed.

### CODE REVIEW

File: weavetest/auth.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	return a.Signers
+}
+
+func (a *Auth) HasAddress(ctx weave.Context, addr weave.Address) bool {
+	for _, s := range a.Signers {
+		if addr.Equals(s.Address()) {
+			return true
+		}
+	}
+	return addr.Equals(a.Signer.Address())
```
Comment (ethanfrey) [inline weavetest/auth.go]: Why separate here? What about range over `a.GetConditions()`

I dislike relying on the handling of `nil.Address`

File: weavetest/auth.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// CtxAuth is a mock implementing x.Authenticator interface.
+//
+// This implementation is using context to store and retrieve permissions.
+type CtxAuth struct {
```
Comment (ethanfrey) [inline weavetest/auth.go]: nice doc strings, good useage (same as original?)

File: weavetest/decorators.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+func (d *Decorator) DeliverCallCount() int {
+	return d.deliverCall
+}
+
+func (d *Decorator) CallCount() int {
+	return d.checkCall + d.deliverCall
+}
+
+func Decorate(h weave.Handler, d weave.Decorator) weave.Handler {
```
Comment (ethanfrey) [inline weavetest/decorators.go]: Looks a lot like app.ChainDecorators.

Not sure if this duplication is good or bad

File: weavetest/doc.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/*
+Package weavetest provides mocks and functions that makes testing weave
+functionality easier.
+
+Before adding a new code to this package make sure that it is used in at least
+two packages. Mocks and helpers that are utilized by only one package should be
```
Comment (ethanfrey) [inline weavetest/doc.go]: Ahhh... this explains some of the questions before about mocks going there.

Great to add these doc strings

File: weavetest/handlers.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package weavetest
+
+import "github.com/iov-one/weave"
+
+// Handler implements a mock of weave.Handler
+//
+// Use this handler in your tests. Set XxxResult and XxxErr to control what Xxx
+// method call returns. Each method call is counted.
+type Handler struct {
```
Comment (ethanfrey) [inline weavetest/handlers.go]: So nice and clean

File: weavetest/tx.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	Err error
+}
+
+var _ weave.Tx = (*Tx)(nil)
+
+func (tx *Tx) GetMsg() (weave.Msg, error) {
+	return tx.Msg, tx.Err
+}
+
+func (tx *Tx) Unmarshal([]byte) error {
```
Comment (ethanfrey) [inline weavetest/tx.go]: Given Msg has Marshal/Unmarshal, probably easy to add them here as well, right?
If that was useful....

File: x/cash/controller_test.go [hunk 7]
```diff
@@ -43,12 +43,8 @@ type checkCmd struct {
 }
 
 func TestIssueCoins(t *testing.T) {
-	var helpers x.TestHelpers
-
-	_, perm := helpers.MakeKey()
-	_, perm2 := helpers.MakeKey()
-	addr := perm.Address()
-	addr2 := perm2.Address()
+	addr1 := weavetest.NewCondition().Address()
```
Comment (ethanfrey) [inline x/cash/controller_test.go]: Definitely nicer

File: weavetest/handlers.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	return h.checkCall
+}
+
+func (h *Handler) DeliverCallCount() int {
+	return h.deliverCall
+}
+
+func (h *Handler) CallCount() int {
+	return h.checkCall + h.deliverCall
+}
```
Comment (ethanfrey) [inline weavetest/handlers.go]: Hmmm... where are the rest.

It seems counting, errors handler, tag handler, and more are all covered here.

But WriteHandler, PanicHandler?

File: weavetest/auth.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/iov-one/weave"
+)
+
+// Auth is a mock implementing x.Authenticator interface.
+//
+// This structure authenticates any of referenced conditions.
+// You can use either Signer or Signers (or both) attributes to reference
+// conditions. This is for the convinience and each time all signers
+// (regardless which attribute) are considered.
+type Auth struct {
```
Comment (ethanfrey) [inline weavetest/auth.go]: Shouldn't there be a few tests in this package?
At least for the auth stuff and decorator wrapping

File: app/chain_test.go [hunk 10]
```diff
@@ -4,55 +4,78 @@ import (
...
+
 	_, err = stack.Deliver(ctx, nil, nil)
 	assert.Error(t, err)
+	assert.Equal(t, 2, c1.DeliverCallCount())
+	assert.Equal(t, 2, c2.DeliverCallCount())
+	assert.Equal(t, 1, c3.DeliverCallCount())
+	assert.Equal(t, 1, h.DeliverCallCount())
+}
+
+type panicAtHeightDecorator int64
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply app/chain_test.go]: okay... so this was only used in this one package. hmm....

File: x/utils/key_tagger_test.go [hunk 11]
```diff
@@ -131,3 +133,56 @@ func TestKeyTagger(t *testing.T) {
...
+		store.Set(d.key, d.value)
+	}
+	res, err := next.Deliver(ctx, store, tx)
+	if d.after && err == nil {
+		store.Set(d.key, d.value)
+	}
+	return res, err
+}
+
+// tagHandler writes the key, value pair and returns the error (may be nil)
```
Comment (ethanfrey) [inline x/utils/key_tagger_test.go]: Can't you implement this with a properly constructed 
`weavetest.Handler{DeliverResult: weave.DeliverResult{Tags:  common.KVPairs{{Key: h.key, Value: h.value}}}}`

sure, reads ugly, but maybe a helper function here rather than writing a struct?

**Comment by ethanfrey** [comment]
One rebase, a little cleanup, and then good to go.

:mario:

File: app/chain_test.go [hunk 12]
```diff
@@ -4,55 +4,78 @@ import (
 	"context"
 	"testing"
 
-	"github.com/stretchr/testify/assert"
-
 	"github.com/iov-one/weave"
-	"github.com/iov-one/weave/x"
+	"github.com/iov-one/weave/weavetest"
 	"github.com/iov-one/weave/x/utils"
+	"github.com/stretchr/testify/assert"
```
[the remark this answers is not in the document]
Comment (husio) [reply app/chain_test.go]: We don't have control over this. I am using [`goimports`](https://godoc.org/golang.org/x/tools/cmd/goimports) which will rewrite imports and sort them into two groups. :robot:

File: app/chain_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (husio) [reply app/chain_test.go]: I have moved all code that was used in only one package to that package. I think it is ok to write custom helpers if you need them. They become weavetest once at least two package would benefit from them.

File: weavetest/auth.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	// considered.
+	Signer weave.Condition
+
+	// Signers represents an authentication of multiple signers.
+	Signers []weave.Condition
+}
+
+func (a *Auth) GetConditions(weave.Context) []weave.Condition {
+	if a.Signer != nil {
+		return append(a.Signers, a.Signer)
```
[the remark this answers is not in the document]
Comment (husio) [reply weavetest/auth.go]: Yes. You should not modify this array once created :smirk:

Should I change it to use a copy of `a.Signers` instead?

File: weavetest/auth.go (hunk 1 above)
Comment (husio) [reply weavetest/auth.go]: Existence of two signer attributes on this structure - `Signer` and `Signers` is for convenience only. I have created  this structure with only `Signers` at first but then if you want to create an authenticator for only one signer you had to write 
```go
auth := &weavetest.Auth{Signers: []weave.Condition{signer}}
```
In comparison, having a `Signer` attribute you are saving yourself typing of 20 characters!! :joy_cat: 
```go
auth := &weavetest.Auth{Signer: signer}
```
Having two signer references means that you can use both of them at the same time as well.
```go
auth := &weavetest.Auth{
    Signers: signersGroup,
    Signer: signer,
}
```
---

I can change the last line of the test to 
```go
if a.Signer == nil { 
    return false 
} 
return addr.Equals(a.Signer.Address())
```
What do you think?

File: weavetest/decorators.go (hunk 3 above)
Comment (husio) [reply weavetest/decorators.go]: Good catch! I did not notice that.

It was called [`Wrap`](https://github.com/iov-one/weave/blob/master/x/helpers.go#L83) before so I have ported it. Should I replace it? I can see both pros and cons.

File: weavetest/doc.go (hunk 4 above)
Comment (husio) [reply weavetest/doc.go]: I belong to the [`WET`](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself#DRY_vs_WET_solutions) church! :church: :sweat_drops: :smiling_imp:

File: weavetest/tx.go (hunk 6 above)
Comment (husio) [reply weavetest/tx.go]: Good catch! Addded in https://github.com/iov-one/weave/pull/386/commits/94633ea2804df58b6e31b7e9a42eb55c5a3aec3c

File: weavetest/handlers.go (hunk 8 above)
Comment (husio) [reply weavetest/handlers.go]: Those are moved to the package that was using them. They were providing a very specific functionality (ie testing transaction) that only one package needed. 

If used once, code is moved to its package. https://github.com/iov-one/weave/pull/386#discussion_r263524416

File: x/utils/key_tagger_test.go (hunk 11 above)
Comment (husio) [reply x/utils/key_tagger_test.go]: I have copied this without reading the implementation :see_no_evil: I will look into this and update if possible.

File: weavetest/auth.go (hunk 2 above)
Comment (husio) [reply weavetest/auth.go]: New docs, implementation the same as before (`Key` type was changed to `string` though).

File: weavetest/auth.go (hunk 1 above)
Comment (ethanfrey) [reply weavetest/auth.go]: Sure.
I very much like the two members, they make it easier to use. 
I just wanted to avoid the comparison to nil

File: weavetest/decorators.go (hunk 3 above)
Comment (ethanfrey) [reply weavetest/decorators.go]: I guess it is fine.
I don't want dependencies between weavetest and app also
Maybe leave a comment about the other version for clarity

**Comment by ethanfrey** [comment]
Good responses. 
Let me know if you need another pass, but I think you got it all.

File: weavetest/auth.go (hunk 9 above)
Comment (husio) [reply weavetest/auth.go]: Done in https://github.com/iov-one/weave/pull/386/commits/137fdf5f6f64ef80759e4ec45be5c2981212964a

File: app/chain_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply app/chain_test.go]: That explains this

File: app/chain_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply app/chain_test.go]: Yeah, got this once i saw the godocs

File: weavetest/auth_test.go [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+		t.Fatalf("unexpected conditions")
+	}
+
+	for i, c := range conds {
+		if !a.HasAddress(nil, c.Address()) {
+			t.Errorf("condition %d (%s) address should be present", i, c)
+		}
+	}
+
+	if a.HasAddress(nil, NewCondition().Address()) {
```
Comment (ethanfrey) [inline weavetest/auth_test.go]: :smile:

Comment (ethanfrey) [reply weavetest/auth_test.go]: What about `nil, nil`. I think this would have matched before
